conda: init at miniconda3 4.3.31

bhipple opened the pull request:
###### Motivation for this change
This takes the work done in PR #26443, which itself was a derivative work from
PR #26245, and adds a few improvements, including:
- Upgrading to the latest miniconda3 installer script
- Adding meta information for licenses, homepage, maintainers
- Removing unnecessary emacs and git packages
- Compact the expressions and docs a little
- Rename the FHS script to conda-shell, to more consistently match conda-install and also be more easily discoverable with tab-complete
- More consistently naming the installer miniconda3, in case we someday want to support miniconda.

On the last point, I don't think we really need to give a miniconda2 option, or
at least it isn't a high priority; conda itself can install and manage conda
environments and pythons, so AFAICT a miniconda2 install would just save a small
amount of download bandwidth when someone who wants exclusively python2 adds a
python2 environment.

CC @fridh @jluttine @layus 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

FRidh committed:
conda: init at miniconda3 4.3.31

bhipple commented:
@GrahamcOfBorg build conda

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34872&attempt_id=9d8924fb-7ab1-489d-a80a-be230f64607c)

<details><summary>Partial log (click to expand)</summary><p>

```
collision between `/nix/store/kzgnkvjy768hd8k3q0p61zchhv17kg3p-gcc-6.4.0-lib/lib/libgcc_s.so.1' and `/nix/store/fij62nvzlwdgih6hgdahv9mxrlzy7kna-glibc-2.26-131/lib/libgcc_s.so.1'
collision between `/nix/store/kzgnkvjy768hd8k3q0p61zchhv17kg3p-gcc-6.4.0-lib/lib64/libgcc_s.so' and `/nix/store/fij62nvzlwdgih6hgdahv9mxrlzy7kna-glibc-2.26-131/lib64/libgcc_s.so'
collision between `/nix/store/kzgnkvjy768hd8k3q0p61zchhv17kg3p-gcc-6.4.0-lib/lib64/libgcc_s.so.1' and `/nix/store/fij62nvzlwdgih6hgdahv9mxrlzy7kna-glibc-2.26-131/lib64/libgcc_s.so.1'
collision between `/nix/store/zhx0mndbdj0191b8ia2vyp85y36pwdkq-shadow-4.5/etc/login.defs' and `/nix/store/h7n3f3mbq5qdzcpvn7slidhi7xrqmhrp-conda-shell-chrootenv-etc/etc/login.defs'
collision between `/nix/store/zhx0mndbdj0191b8ia2vyp85y36pwdkq-shadow-4.5/etc/pam.d' and `/nix/store/h7n3f3mbq5qdzcpvn7slidhi7xrqmhrp-conda-shell-chrootenv-etc/etc/pam.d'
created 1148 symlinks in user environment
building '/nix/store/94lrd8k4qvncb462idygximfp5n011zv-conda-shell-fhs.drv'...
building '/nix/store/k0rai71m4kq53b5kg77wz6cf264d40mc-conda-shell-init.drv'...
building '/nix/store/slcx530hyiaig829r3hph7sskyj407kh-conda-shell.drv'...
/nix/store/5h6kg5x1644si26rq4l2ybyv8xma140q-conda-shell
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34872&attempt_id=a07928e4-653c-4370-952f-23fa2a79be58)

<details><summary>Partial log (click to expand)</summary><p>

```
collision between `/nix/store/kzqrxkxs6lcbhxlalcim7zz6xay0i2f2-gcc-6.4.0-lib/lib/libgcc_s.so.1' and `/nix/store/diqn7sliwgm97bv83aklxxmpjdz4n2sj-glibc-multi-2.26-131/lib/libgcc_s.so.1'
collision between `/nix/store/kzqrxkxs6lcbhxlalcim7zz6xay0i2f2-gcc-6.4.0-lib/lib64/libgcc_s.so' and `/nix/store/diqn7sliwgm97bv83aklxxmpjdz4n2sj-glibc-multi-2.26-131/lib64/libgcc_s.so'
collision between `/nix/store/kzqrxkxs6lcbhxlalcim7zz6xay0i2f2-gcc-6.4.0-lib/lib64/libgcc_s.so.1' and `/nix/store/diqn7sliwgm97bv83aklxxmpjdz4n2sj-glibc-multi-2.26-131/lib64/libgcc_s.so.1'
collision between `/nix/store/w2kj9zddzmfsjmkfnbnypblm54hdi9d7-shadow-4.5/etc/login.defs' and `/nix/store/9i4m29p8nslxjq7rqbbvb1j0y7s840zr-conda-shell-chrootenv-etc/etc/login.defs'
collision between `/nix/store/w2kj9zddzmfsjmkfnbnypblm54hdi9d7-shadow-4.5/etc/pam.d' and `/nix/store/9i4m29p8nslxjq7rqbbvb1j0y7s840zr-conda-shell-chrootenv-etc/etc/pam.d'
created 1162 symlinks in user environment
building '/nix/store/5alv2fg6cbii6axl5dm7vmzc04w5wvjf-conda-shell-fhs.drv'...
building '/nix/store/7sjx0l2j0hq21ygirw58qdb4w22ahg8a-conda-shell-init.drv'...
building '/nix/store/hpgi0m4lr71gq2xs0v7akah2w8v0swf2-conda-shell.drv'...
/nix/store/g8qikismr6h66hn094c1ciirdcbvjgw6-conda-shell
```
</p></details>

jluttine commented:
Thanks! I didn't test it but took a quick look and looks good to me. Great that this is being worked on!

Just some general questions:

- If one wanted to add some packages to the FHS conda environment, how is that done? With overlays? Or what is (or would be) the most convenient way to support that?

- Can I easily define different conda environments for my projects? So that I add `shell.nix` to each of my project directory and those files can add whatever system packages the project needs to the conda FHS environment. Then, instead of running `conda-shell`, I'd run `nix-shell` in the project directory. Should some instructions on this be added?

I'm relatively new to nix so I suppose these are quite easy to achieve, but just wondering what is the "correct" way.

FRidh commented on pkgs/development/interpreters/conda/default.nix:
```diff
@@ -0,0 +1,98 @@
+{ lib
+, pkgs
```
Not `pkgs` but individual packages

FRidh commented on pkgs/development/interpreters/conda/default.nix:
```diff
@@ -0,0 +1,98 @@
+# $ conda-shell
+# $ conda install spyder
+let
+  minicondaInstaller = pkgs.stdenv.mkDerivation rec {
```
I think it makes more sense to use `runCommand` here

FRidh commented on pkgs/development/interpreters/conda/default.nix:
```diff
@@ -0,0 +1,98 @@
+    name = "conda-shell";
+    targetPkgs = pkgs: (
+      with pkgs; [
+        conda
```
There should likely be a way to extend the packages here.

FRidh commented on pkgs/development/interpreters/conda/default.nix:
```diff
@@ -0,0 +1,98 @@
+      homepage = https://conda.io/;
+      platforms = lib.platforms.linux;
+      license = lib.licenses.bsd3;
+      maintainers = with lib.maintainers; [ jluttine fridh bhipple ];
```
Please remove me from the maintainers. I don't intend to use it.

FRidh reviewed (commented):
Thanks for taking this on!

bhipple committed:
Code review updates

- Overlay-friendly extraPkgs list for FHS customization
- Update maintainers
- Simplify installer script download significantly
- Explicitly mention packages used, instead of pkgs

bhipple commented on pkgs/development/interpreters/conda/default.nix:
```diff
@@ -0,0 +1,71 @@
+# Conda manages most pkgs itself, but expects a few to be on the system.
+, condaDeps ? [ stdenv.cc xorg.libSM xorg.libICE xorg.libXrender libselinux ]
+# Any extra nixpkgs you'd like available in the FHS env for Conda to use
+, extraPkgs ? [ ]
```
This can be extended with an overlay as such:
```
$ cat ~/.config/nixpkgs/overlays/conda.nix                                                                                 ~/src/nixpkgs
self: super:
{
  conda = super.conda.override {
    extraPkgs = [ self.sssd self.firefox ];
  };
}

$ conda-shell
$ ls -l /usr/bin/firefox
lrwxrwxrwx 1 nobody nogroup 78 Dec 31  1969 /usr/bin/firefox -> /nix/store/a258dljnrj0chhnb6dpir2106yimfkb9-conda-shell-usr-target/bin/firefox

$ ls -l /usr/lib/sssd/libsss_krb5.so
lrwxrwxrwx 1 nobody nogroup 90 Dec 31  1969 /usr/lib/sssd/libsss_krb5.so -> /nix/store/a258dljnrj0chhnb6dpir2106yimfkb9-conda-shell-usr-target/lib/sssd/libsss_krb5.so
```

bhipple commented on pkgs/development/interpreters/conda/default.nix:
```diff
@@ -0,0 +1,71 @@
+  src = fetchurl {
+      url = "https://repo.continuum.io/miniconda/Miniconda3-${version}-Linux-x86_64.sh";
+      sha256 = "1rklq81s9v7xz1q0ha99w2sl6kyc5vhk6b21cza0jr3b8cgz0lam";
+  };
```
Thanks for the tip to get rid of `mkDerivation` @FRidh; this is WAY simpler!

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34872&attempt_id=d19732ef-8aff-44c4-9786-45210633e1a3)

<details><summary>Partial log (click to expand)</summary><p>

```
collision between `/nix/store/kzqrxkxs6lcbhxlalcim7zz6xay0i2f2-gcc-6.4.0-lib/lib64/libgcc_s.so.1' and `/nix/store/diqn7sliwgm97bv83aklxxmpjdz4n2sj-glibc-multi-2.26-131/lib64/libgcc_s.so.1'
collision between `/nix/store/kzqrxkxs6lcbhxlalcim7zz6xay0i2f2-gcc-6.4.0-lib/lib/libgcc_s.so' and `/nix/store/diqn7sliwgm97bv83aklxxmpjdz4n2sj-glibc-multi-2.26-131/lib/libgcc_s.so'
collision between `/nix/store/kzqrxkxs6lcbhxlalcim7zz6xay0i2f2-gcc-6.4.0-lib/lib/libgcc_s.so.1' and `/nix/store/diqn7sliwgm97bv83aklxxmpjdz4n2sj-glibc-multi-2.26-131/lib/libgcc_s.so.1'
collision between `/nix/store/w2kj9zddzmfsjmkfnbnypblm54hdi9d7-shadow-4.5/etc/login.defs' and `/nix/store/9i4m29p8nslxjq7rqbbvb1j0y7s840zr-conda-shell-chrootenv-etc/etc/login.defs'
collision between `/nix/store/w2kj9zddzmfsjmkfnbnypblm54hdi9d7-shadow-4.5/etc/pam.d' and `/nix/store/9i4m29p8nslxjq7rqbbvb1j0y7s840zr-conda-shell-chrootenv-etc/etc/pam.d'
created 1163 symlinks in user environment
building path(s) ‘/nix/store/0qhpl5d4009xfgncjs3a5fyhlb3zc8rx-conda-shell-fhs’
building path(s) ‘/nix/store/r25zy2hzw2za42idjbzwvp1yxmkc87bg-conda-shell-init’
building path(s) ‘/nix/store/qpzbjv2yclh49gfmfsnd1qmnqf6k3ysf-conda-shell’
/nix/store/qpzbjv2yclh49gfmfsnd1qmnqf6k3ysf-conda-shell
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34872&attempt_id=62968763-a5c5-4175-b140-df6e96f5d3ad)

<details><summary>Partial log (click to expand)</summary><p>

```
collision between `/nix/store/kzgnkvjy768hd8k3q0p61zchhv17kg3p-gcc-6.4.0-lib/lib/libgcc_s.so.1' and `/nix/store/fij62nvzlwdgih6hgdahv9mxrlzy7kna-glibc-2.26-131/lib/libgcc_s.so.1'
collision between `/nix/store/kzgnkvjy768hd8k3q0p61zchhv17kg3p-gcc-6.4.0-lib/lib64/libgcc_s.so' and `/nix/store/fij62nvzlwdgih6hgdahv9mxrlzy7kna-glibc-2.26-131/lib64/libgcc_s.so'
collision between `/nix/store/kzgnkvjy768hd8k3q0p61zchhv17kg3p-gcc-6.4.0-lib/lib64/libgcc_s.so.1' and `/nix/store/fij62nvzlwdgih6hgdahv9mxrlzy7kna-glibc-2.26-131/lib64/libgcc_s.so.1'
collision between `/nix/store/zhx0mndbdj0191b8ia2vyp85y36pwdkq-shadow-4.5/etc/login.defs' and `/nix/store/h7n3f3mbq5qdzcpvn7slidhi7xrqmhrp-conda-shell-chrootenv-etc/etc/login.defs'
collision between `/nix/store/zhx0mndbdj0191b8ia2vyp85y36pwdkq-shadow-4.5/etc/pam.d' and `/nix/store/h7n3f3mbq5qdzcpvn7slidhi7xrqmhrp-conda-shell-chrootenv-etc/etc/pam.d'
created 1149 symlinks in user environment
building '/nix/store/qzifyl0rmik5hh9qzjs4wihqgdh461j7-conda-shell-fhs.drv'...
building '/nix/store/w33ybqgd797fb684dsw7ixdjzcc71a3q-conda-shell-init.drv'...
building '/nix/store/skn6bz6hxsxvfa07ikb28zqsp3cirxvv-conda-shell.drv'...
/nix/store/lhb4l1ghz0qjn0x3sx29lal6jcz4ip17-conda-shell
```
</p></details>

bhipple commented:
@jluttine regarding your second question, conda has full support for managing its own environments in `~/.conda/` and switching seamlessly between them; since it saves those envs in its state directory they won't be lost between nix sessions.  I don't think we want to interfere too much with the actual package management and environment management of conda itself, since it's already reasonably good at it, but I could imagine a configurable shellHook that runs `source activate <conda env name>` for you automatically after running `conda-shell`.  Conda is changing the way environments are activated and toggled in the upcoming major version, though, so we might want to hold off on implementing that wrapper feature for the time being.

FRidh commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6943,6 +6943,8 @@ with pkgs;
   };
   purePackages = recurseIntoAttrs (callPackage ./pure-packages.nix {});
 
+  conda = callPackage ../development/interpreters/conda { };
```
I think we should actually move the expression to `pkgs/tools/package-management/`

bhipple committed:
Move conda to pkgs/tools/package-management/

FRidh commented:
In general it is preferred when commits are "ready to merge" and that would mean that unnecessary commits have been squashed.

bhipple commented:
Looks like you used the github squash-merge option, so we get a nice clean history without the reviewer having to double check that the claimed "no-op" manual rebase and squash is actually so :+1:

I see some other mergers don't use it, though -- it might be worth noting something in https://nixos.org/nixpkgs/manual/#idm140737316961072 to encourage it when doing so when the individual commits don't contain particularly useful information.

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34872&attempt_id=0a646c52-c2b9-4705-a84b-3e68e03f3a2f)

<details><summary>Partial log (click to expand)</summary><p>

```
collision between `/nix/store/q3d0qg3rsdjkfw7zbxa5cd71rfsbspl0-gcc-6.4.0-lib/lib64/libgcc_s.so' and `/nix/store/6aqa4qjkx4q1zxinkrp14r2k1k13ph8p-glibc-2.26-131/lib64/libgcc_s.so'
collision between `/nix/store/q3d0qg3rsdjkfw7zbxa5cd71rfsbspl0-gcc-6.4.0-lib/lib/libgcc_s.so.1' and `/nix/store/6aqa4qjkx4q1zxinkrp14r2k1k13ph8p-glibc-2.26-131/lib/libgcc_s.so.1'
collision between `/nix/store/q3d0qg3rsdjkfw7zbxa5cd71rfsbspl0-gcc-6.4.0-lib/lib/libgcc_s.so' and `/nix/store/6aqa4qjkx4q1zxinkrp14r2k1k13ph8p-glibc-2.26-131/lib/libgcc_s.so'
collision between `/nix/store/hl22lhhz9bfin4bp3pjq6i1xmw8acr6y-shadow-4.5/etc/pam.d' and `/nix/store/l463zacwgdwc4nc2ylvdqfklwys23pcc-conda-shell-chrootenv-etc/etc/pam.d'
collision between `/nix/store/hl22lhhz9bfin4bp3pjq6i1xmw8acr6y-shadow-4.5/etc/login.defs' and `/nix/store/l463zacwgdwc4nc2ylvdqfklwys23pcc-conda-shell-chrootenv-etc/etc/login.defs'
created 1149 symlinks in user environment
building '/nix/store/09jsyfrxvr1ilhr0bx3p8mqv6i9cami8-conda-shell-fhs.drv'...
building '/nix/store/51a8gdcpi9b4w0x347dlryk5b0n2f8kr-conda-shell-init.drv'...
building '/nix/store/basr4zvk7yqmr1kagsbm90k9pqb9gyvh-conda-shell.drv'...
/nix/store/a9y08alhd6ji31rqjyr8a2p5f3nl1si2-conda-shell
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34872&attempt_id=3e7ff016-f29f-4bc1-b518-bea08dbee7c8)

<details><summary>Partial log (click to expand)</summary><p>

```
collision between `/nix/store/003d91yhs72g120rkshp2d2hqsvi2368-gcc-6.4.0-lib/lib64/libgcc_s.so.1' and `/nix/store/jz04c5060whhj9my2qvj9hkh39qjf1q6-glibc-multi-2.26-131/lib64/libgcc_s.so.1'
collision between `/nix/store/003d91yhs72g120rkshp2d2hqsvi2368-gcc-6.4.0-lib/lib/libgcc_s.so' and `/nix/store/jz04c5060whhj9my2qvj9hkh39qjf1q6-glibc-multi-2.26-131/lib/libgcc_s.so'
collision between `/nix/store/003d91yhs72g120rkshp2d2hqsvi2368-gcc-6.4.0-lib/lib/libgcc_s.so.1' and `/nix/store/jz04c5060whhj9my2qvj9hkh39qjf1q6-glibc-multi-2.26-131/lib/libgcc_s.so.1'
collision between `/nix/store/8rkq0dwlgq7cs8gqkay7h1crwr42ksfv-shadow-4.5/etc/login.defs' and `/nix/store/0hpxc2mas6zs9zjzwsr10kkldf7cjv90-conda-shell-chrootenv-etc/etc/login.defs'
collision between `/nix/store/8rkq0dwlgq7cs8gqkay7h1crwr42ksfv-shadow-4.5/etc/pam.d' and `/nix/store/0hpxc2mas6zs9zjzwsr10kkldf7cjv90-conda-shell-chrootenv-etc/etc/pam.d'
created 1163 symlinks in user environment
building path(s) ‘/nix/store/cis6hnx4rybdwhr8hff3ky53h54j2vgf-conda-shell-fhs’
building path(s) ‘/nix/store/b855pc8fvya2kpy23067cnpqz71dk1hj-conda-shell-init’
building path(s) ‘/nix/store/imfgcmhkjib8jdppcb4v27l8i3kdsbwr-conda-shell’
/nix/store/imfgcmhkjib8jdppcb4v27l8i3kdsbwr-conda-shell
```
</p></details>

FRidh commented:
I know not everyone likes squashing as a maintainer, as that means purging the contributors history (although of course the original commits are still existing on the original branch).
